Very cool. Build failure real:
```
======================================================================
ERROR: test_case_has_transactions_since_sync (corehq.form_processor.tests.test_case_dbaccessor.CaseAccessorTestsSQL)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/corehq/form_processor/tests/test_case_dbaccessor.py", line 29, in tearDown
    FormProcessorTestUtils.delete_all_sql_forms(DOMAIN)
  File "/mnt/corehq/form_processor/tests/utils.py", line 91, in delete_all_sql_forms
    FormAccessorSQL.delete_all_forms(domain, user_id)
  File "/mnt/corehq/util/test_utils.py", line 32, in inner
    return fn(*args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/db/transaction.py", line 394, in inner
    return func(*args, **kwargs)
  File "/mnt/corehq/form_processor/backends/sql/dbaccessors.py", line 263, in delete_all_forms
    cursor.execute('SELECT delete_all_forms(%s, %s)', [domain, user_id])
  File "/vendor/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/vendor/lib/python2.7/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/vendor/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
InternalError: PL/Proxy function public.delete_all_forms(2): [test_test_commcarehq_p1] PQconnectPoll: FATAL:  database "test_test_commcarehq_p1" does not exist
```

--- a/corehq/sql_db/config.py
+++ b/corehq/sql_db/config.py
@@ -42,15 +42,8 @@ class DbShard(object):
         self.shard_id = shard_id
         self.django_dbname = django_dbname
 
-    def get_db_config(self):
-        from django.db.backends.creation import TEST_DATABASE_PREFIX
-        db_config = settings.DATABASES[self.django_dbname].copy()
-        if settings.UNIT_TESTING:
-            db_config['NAME'] = TEST_DATABASE_PREFIX + db_config['NAME']
-        return db_config
-
     def to_shard_meta(self, host_map):
-        config = self.get_db_config()
+        config = settings.DATABASES[self.django_dbname]
         host = host_map.get(config['HOST'], config['HOST'])
         return ShardMeta(
             id=self.shard_id,

--- a/corehq/sql_db/tests/test_partition_config.py
+++ b/corehq/sql_db/tests/test_partition_config.py
@@ -68,10 +68,10 @@ class TestPartitionConfig(SimpleTestCase):
         config = PartitionConfig()
         shards = config.get_shards()
         self.assertEquals(shards, [
-            ShardMeta(id=0, dbname='test_db1', host='hqdb1', port=5432),
-            ShardMeta(id=1, dbname='test_db1', host='hqdb1', port=5432),
-            ShardMeta(id=2, dbname='test_db2', host='hqdb2', port=5432),
-            ShardMeta(id=3, dbname='test_db2', host='hqdb2', port=5432),
+            ShardMeta(id=0, dbname='db1', host='hqdb1', port=5432),
+            ShardMeta(id=1, dbname='db1', host='hqdb1', port=5432),
+            ShardMeta(id=2, dbname='db2', host='hqdb2', port=5432),
+            ShardMeta(id=3, dbname='db2', host='hqdb2', port=5432),
         ])
 
     @override_settings(PARTITION_DATABASE_CONFIG=TEST_PARTITION_CONFIG_HOST_MAP)
@@ -79,8 +79,8 @@ class TestPartitionConfig(SimpleTestCase):
         config = PartitionConfig()
         shards = config.get_shards()
         self.assertEquals(shards, [
-            ShardMeta(id=0, dbname='test_db1', host='localhost', port=5432),
-            ShardMeta(id=1, dbname='test_db2', host='hqdb2', port=5432),
+            ShardMeta(id=0, dbname='db1', host='localhost', port=5432),
+            ShardMeta(id=1, dbname='db2', host='hqdb2', port=5432),
         ])
 
     @override_settings(PARTITION_DATABASE_CONFIG=INVALID_SHARD_RANGE_START)
@@ -105,8 +105,8 @@ class PlProxyTests(SimpleTestCase):
 
     def test_get_server_option_string(self):
         self.assertEqual(
-            "p0 'dbname=test_db1 host=hqdb0 port=5432'",
-            ShardMeta(id=0, dbname='test_db1', host='hqdb0', port=5432).get_server_option_string()
+            "p0 'dbname=db1 host=hqdb0 port=5432'",
+            ShardMeta(id=0, dbname='db1', host='hqdb0', port=5432).get_server_option_string()
         )
 
     def test_parse_existing_shard(self):

--- a/corehq/tests/nose.py
+++ b/corehq/tests/nose.py
@@ -284,8 +284,11 @@ class HqdbContext(DatabaseContext):
     """
 
     def __init__(self, tests, runner):
-        self.optimize = os.environ.get("REUSE_DB", "optimize") == "optimize"
-        if self.optimize:
+        reuse_db = os.environ.get("REUSE_DB")
+        self.skip_setup_for_reuse_db = reuse_db and reuse_db not in ["reset", "optimize"]
+        self.skip_teardown_for_reuse_db = reuse_db and reuse_db != "teardown"
+        self.optimize_migrations = reuse_db in [None, "optimize"]
+        if self.optimize_migrations:
             self.test_labels = AppLabelsPlugin.get_test_labels(tests)
         super(HqdbContext, self).__init__(tests, runner)
 
@@ -303,7 +306,7 @@ class HqdbContext(DatabaseContext):
         if self.should_skip_test_setup():
             return
 
-        if self.optimize:
+        if self.optimize_migrations:
             self.optimizer = optimize_apps_for_test_labels(self.test_labels)
             self.optimizer.__enter__()
 
@@ -317,7 +320,7 @@ class HqdbContext(DatabaseContext):
             databases = {app_name: uri for app_name, uri in databases}
         self.apps = [self.verify_test_db(*item) for item in databases.items()]
 
-        if os.environ.get("REUSE_DB") not in ["reset", "optimize"]:
+        if self.skip_setup_for_reuse_db:
             from django.db import connections
             old_names = []
             for connection in connections.all():
@@ -340,11 +343,10 @@ class HqdbContext(DatabaseContext):
             return
 
         self.blob_db.close()
-        if self.optimize:
+        if self.optimize_migrations:
             self.optimizer.__exit__(None, None, None)
 
-        if os.environ.get("REUSE_DB") != "teardown":
-            # skip remaining teardown
+        if self.skip_teardown_for_reuse_db:
             return
 
         # delete couch databases

